tests/endpoints-2.0/endpoints-integration.spec.ts

```diff
@@ -1,136 +1,98 @@
-import * as fs from "fs";
-import * as path from "path";
+import { EndpointV2 } from "@smithy/types";
+import { resolveEndpoint, EndpointParams } from "@smithy/util-endpoints";
+import { existsSync, readdirSync } from "fs";
+import { join } from "path";
 
-import { EndpointExpectation, EndpointTestCase, ErrorExpectation, ServiceNamespace } from "./integration-test-types";
-
-const clientList: string[] = [];
-const root = path.join(__dirname, "..", "..");
-const clients = fs.readdirSync(path.join(root, "clients"));
-clientList.push(...clients);
+import { EndpointExpectation, ServiceModel, ServiceNamespace } from "./integration-test-types";
 
 describe("client list", () => {
+  const root = join(__dirname, "..", "..");
+  const clientPackageNameList = readdirSync(join(root, "clients"));
+
   it("should be at least 300 clients", () => {
-    expect(clientList.length).toBeGreaterThan(300);
+    expect(clientPackageNameList.length).toBeGreaterThan(300);
   });
-});
 
-for (const client of clientList) {
-  const serviceName = client.slice(7);
+  describe.each(clientPackageNameList)(`%s endpoint test cases`, (clientPackageName) => {
+    const serviceName = clientPackageName.slice(7);
 
-  let defaultEndpointResolver: any;
-  let namespace: any;
-  let model: any;
+    // since client package name list is populated from clients folder, we know it exists.
+    const namespace = require(`@aws-sdk/${clientPackageName}`);
+    const modelPath = join(root, "codegen", "sdk-codegen", "aws-models", serviceName + ".json");
 
-  // this may also work with dynamic async import() in a beforeAll() block,
-  // but needs more effort than using synchronous require().
-  try {
-    defaultEndpointResolver =
-      require(`@aws-sdk/client-${serviceName}/src/endpoint/endpointResolver`).defaultEndpointResolver;
-    namespace = require(`@aws-sdk/client-${serviceName}`);
-    model = require(path.join(root, "codegen", "sdk-codegen", "aws-models", serviceName + ".json"));
-  } catch (e) {
-    defaultEndpointResolver = null;
-    namespace = null;
-    model = null;
-    if (e.code !== "MODULE_NOT_FOUND") {
-      console.error(e);
-    }
-  }
-
-  describe(`client-${serviceName} endpoint test cases`, () => {
-    if (defaultEndpointResolver && namespace && model) {
-      const [, service] = Object.entries(model.shapes).find(
-        ([k, v]) => typeof v === "object" && v !== null && "type" in v && v.type === "service"
-      ) as any;
-      const [, tests] = Object.entries(service.traits).find(([k, v]) => k === "smithy.rules#endpointTests") as any;
-      if (tests?.testCases) {
-        runTestCases(tests, service, defaultEndpointResolver, "");
-      } else {
-        it.skip("has no test cases", () => {});
+    if (existsSync(modelPath)) {
+      const model = require(modelPath);
+      for (const value of Object.values(model.shapes)) {
+        if (typeof value === "object" && value !== null && "type" in value && value.type === "service") {
```

Comment:
is this equivalent?

```ts
if (value?.type === "service") {
```

Reply:
or 

```ts
const serviceShape = 
  Object.values(model.shapes).find(shape => shape?.type === "service");
```

Reply:
No. TS throws error `Property 'type' does not exist on type 'object'` since type may not exist on value.